src/util/route.js

```diff
@@ -6,13 +6,13 @@ export function isSameRoute (a: Route, b: ?Route): boolean {
   } else if (a.path && b.path) {
     return (
       a.path === b.path &&
-      a.hash === b.hash &&
+      (a.hash || '') === (b.hash || '') &&
```

Comment:
@posva I think I've found the cause: It's in router-link's render function, where it passed a `Location ` object (whose `hash` is optional) directly into isSameRoute:
see https://github.com/vuejs/vue-router/blob/next/src/components/link.js#L31-L33

It seems that `const to` will always get a `path` after `normalizeLocation` : 
https://github.com/vuejs/vue-router/blob/next/src/components/link.js#L24-L31

And it makes `compareTarget` always `to` instead of `resolved`: 
https://github.com/vuejs/vue-router/blob/next/src/components/link.js#L31

I'm surprised that flow didn't warn on implicit type conversion: passing a `Location` object to `isSameRoute`, while it asked for a `Route`